src/components/chatbox/chatbox.tsx

```diff
@@ -0,0 +1,550 @@
+				})
+			}
+			return next
+		})
+	}
+
+	// update component state when text is streamed in from the backend
+	useEffect(() => {
+		if (!streamedText || !messageId || !conversationId) {
+			return
+		}
+		updateAssistantMessageByIds({
+			conversationId,
+			messageId,
+			text: streamedText,
+		})
+	}, [streamedText, messageId, conversationId])
+
+	const handleSubmit = async (e) => {
+		const task = e.currentTarget.task?.value
+		e.preventDefault()
+
+		// Reset previous messages
+		// -- Revisit this when we're ready for multi-message conversations
+		resetMessageList()
+
+		// Clear textarea
+		setUserInput('')
+
+		sendMessage({
+			value: task,
+			// -- uncomment these parameters when we are ready for multi-message conversations
+			// conversationId,
+			// parentMessageId: messageId,
+		})
+
+		// append user message to list
+		appendMessage({ type: 'user', text: task, image: user.image })
+	}
+
+	// Throttle this value to enable uninterrupted smooth scrolling
+	const throttledText = useThrottle(streamedText, 400)
+
+	// scroll text response to bottom as text streams in from the backend
+	const textContentRef = useRef<HTMLDivElement>(null)
+	useEffect(() => {
+		textContentRef.current?.scrollTo({
+			top: textContentRef.current.scrollHeight,
+			behavior: 'smooth',
+		})
+	}, [throttledText])
+
+	// for imperatively submitting the form via textarea-enter-key
+	const formRef = useRef<HTMLFormElement>(null)
+
+	// for conditionally rendering a down arrow
+	const [textContentRef2, textContentScrollBarIsVisible] = useScrollBarVisible()
+
+	// update component state when the mutation fails
+	useEffect(() => {
+		if (errorText) {
+			appendMessage({ type: 'application', text: errorText })
+		}
+	}, [errorText])
+
+	const handleArrowDownClick = () => {
+		textContentRef.current?.scrollTo({
+			top: textContentRef.current.scrollHeight,
+			behavior: 'smooth',
+		})
+	}
+
+	return (
+		<div className={cn(s.chat)}>
+			{messageList.length === 0 ? (
+				<WelcomeMessage />
+			) : (
+				<div
+					ref={mergeRefs([textContentRef, textContentRef2])}
+					className={s.message_list}
+				>
+					{messageList.map((e, i) => {
+						switch (e.type) {
+							// User input
+							case 'user': {
+								return (
+									<UserMessage
+										key={`${e.type}-${i}`}
+										image={e.image}
+										text={e.text}
+									/>
+								)
+							}
+							// Backend AI response
+							case 'assistant': {
+								const shouldShowActions = e.text?.length > 20 && !isLoading
+								return (
+									<AssistantMessage
+										markdown={e.text}
+										key={e.messageId}
+										conversationId={e.conversationId}
+										messageId={e.messageId}
+										showActions={shouldShowActions}
+									/>
+								)
+							}
+							// Application message; likely an error
+							case 'application': {
+								return (
+									<ApplicationMessage key={`${e.type}-${i}`} text={e.text} />
+								)
+							}
+						}
```

Comment:
Can you abstract this to a single component?

Reply:
Hmm, maybe? Though in this case it _feels more simple_ to keep the 3 components separate. The variance in props-interface doesn't feel right to mash into one component.

WDYT?

Reply:
Maybe the suggestion was more like can we abstract this into a single component that still has the switch statement, just to make it easier to read the conditional?

Reply:
Here's the abstraction.

```tsx
<MessageList messages={messageList} />
```

And a visual to show the change in readability

<img width="1324" alt="CleanShot 2023-09-21 at 09 56 02@2x" src="https://github.com/hashicorp/dev-portal/assets/26389321/25bc66c4-2b64-4c99-b5c3-c2a9ad8061d2">

Reply:
Abstracted and improved in a043230c0